c10/core/ConstantSymNodeImpl.cpp

```diff
@@ -1,23 +1,30 @@
 #include <c10/core/ConstantSymNodeImpl.h>
 
 namespace c10 {
-// Temporary hack to avoid having to implement multiple dispatch for now
-// Currently even if we have this method, we still raise an error when we get
-// to SingletonSymNode::eq since comparing with non-singleton is disallowed.
-// However, we may change that behavior in the future.
-template <typename T>
-c10::SymNode ConstantSymNodeImpl<T>::eq(const c10::SymNode& other) {
-  TORCH_INTERNAL_ASSERT(other->singleton_int().has_value());
-  c10::raw::intrusive_ptr::incref(this);
-  return other->eq(c10::intrusive_ptr<ConstantSymNodeImpl<T>>::reclaim(this));
-}
-template <typename T>
-c10::SymNode ConstantSymNodeImpl<T>::ne(const c10::SymNode& other) {
-  TORCH_INTERNAL_ASSERT(other->singleton_int().has_value());
-  c10::raw::intrusive_ptr::incref(this);
-  return other->ne(c10::intrusive_ptr<ConstantSymNodeImpl<T>>::reclaim(this));
-}
+
+// This is used to support the case where the lhs is a constant symnode
+// and the rhs is a singleton symnode. This situation occurs today when we
+// perform a binary op between singleton int and plain int and the
+// singleton promotes the int into a constant symnode. If we'd like to
+// support more combinations in the future, we may need to implement some
+// kind of multiple dispatch.
+#define DEFINE_BINARY_OP(OP)                                           \
+  template <typename T>                                                \
+  c10::SymNode ConstantSymNodeImpl<T>::OP(const c10::SymNode& other) { \
+    TORCH_INTERNAL_ASSERT(other->singleton_int().has_value());         \
+    c10::raw::intrusive_ptr::incref(this);                             \
+    return other->OP(                                                  \
+        c10::intrusive_ptr<ConstantSymNodeImpl<T>>::reclaim(this));    \
```

Comment:
This is old code, but it's a bit weird to me that you incref and then reclaim, I feel like there should be a more direct API for this.